### PR TITLE
Fix bug in a unit test

### PR DESCRIPTION
**Pull request by jdavcs**
There was a bug in a context manager inside a unit test. This caused
nondeterministic behavior of some unit tests after a refactoring.

Bug fixed; added 3 unit tests to test the context manager.

Ping @jmchilton @natefoo

### CODE REVIEW

**Review by natefoo — APPROVED** [review]
Good catch!

File: test/unit/tools/test_tool_deps.py [hunk 1]
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
 
 
 @contextmanager
-def __environ(values, remove=[]):
+def __environ(values, keys_to_remove=[]):
```
Comment (mvdbeek) [inline test/unit/tools/test_tool_deps.py]: I distribute bonus points for eliminating mutable default args, see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments. The way you use it here it's fine,
but better to just say `keys_to_remove=None` and then 
```
if keys_to_remove is None:
    keys_to_remove = []
```
in the function body.

File: test/unit/tools/test_tool_deps.py [hunk 2]
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
...
     """
     Modify the environment for a test, adding/updating values in dict `values` and
     removing any environment variables mentioned in list `remove`.
```
Comment (nsoranzo) [inline test/unit/tools/test_tool_deps.py]:
```suggestion
    removing any environment variables mentioned in list `keys_to_remove`.
```

File: test/unit/tools/test_tool_deps.py [hunk 3]
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
...
+        os.environ must be preserved across calls to __environ
+        when updating existing or adding new key:value pairs
+    """
+    key, val = 'foo_test_678363', '42'
+    os.environ[key] = val
+    new_key, new_val1 = 'foo_test_new', '53'  # add new key + value
+    new_val2 = '43'  # update old key's value
+    to_update = {new_key: new_val1, key: new_val2}
+    with __environ(values=to_update):
+        pass
```
Comment (nsoranzo) [inline test/unit/tools/test_tool_deps.py]: What about testing that `os.environ` has been updated here?

File: test/unit/tools/test_tool_deps.py [hunk 4]
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
...
+    """
+        os.environ must be preserved across calls to __environ
+        when removing key:value pairs
+    """
+    key, val = 'foo_test_678363', '42'
+    os.environ[key] = val
+    new_key, new_val = 'foo_test_new', '53'  # add new key + value
+    to_update = {new_key: new_val}
+    with __environ(values=to_update, keys_to_remove=[key, new_key]):
+        pass
```
Comment (nsoranzo) [inline test/unit/tools/test_tool_deps.py]: Same here.

File: test/unit/tools/test_tool_deps.py (hunk 1 above)
Comment (jdavcs) [reply test/unit/tools/test_tool_deps.py]: @mvdbeek Yes! (and there's no need to create [] in the body: if it's none, we never use it - so a simple check will be sufficient). Thx!

File: test/unit/tools/test_tool_deps.py (hunk 2 above)
Comment (jdavcs) [reply test/unit/tools/test_tool_deps.py]: @nsoranzo Of course!

**Comment by martenson** [comment]
switching to wip until the comments are addressed

File: test/unit/tools/test_tool_deps.py (hunk 4 above)
Comment (jdavcs) [reply test/unit/tools/test_tool_deps.py]: Done!

File: test/unit/tools/test_tool_deps.py (hunk 3 above)
Comment (jdavcs) [reply test/unit/tools/test_tool_deps.py]: Done!

File: test/unit/tools/test_tool_deps.py (hunk 2 above)
Comment (jdavcs) [reply test/unit/tools/test_tool_deps.py]: Done.

File: test/unit/tools/test_tool_deps.py (hunk 1 above)
Comment (jdavcs) [reply test/unit/tools/test_tool_deps.py]: Done.

**Comment by jdavcs** [comment]
I have made several modifications to the context manager + tests (also, see notes in latest commit). As far as I can tell, the main change to the context manager (no more new_keys) shouldn't cause any issues. Please take a look. @nsoranzo @mvdbeek

**Comment by jdavcs** [comment]
E: Unable to locate package libpython3.5-dev??
Can someone rerun the tests, please?

**Comment by jmchilton** [comment]
Looks good to me, that bug with Python 3.5 is fixed upstream so I wouldn't worry about it. Take out of WIP?

**Review by jmchilton — APPROVED** [review]

**Comment by jdavcs** [comment]
@jmchilton thanks! Yes, all good to go then.

**Comment by jdavcs** [comment]
This one is new: OperationalError: (psycopg2.OperationalError) FATAL:  remaining connection slots are reserved for non-replication superuser connections. It's unrelated to this PR (as far as I can judge). Will rerun tests.

**Comment by jdavcs** [comment]
Ready for merging!

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**